Ensure IOPromises end when a request/block ends with CancelContext

theojulienne opened the pull request:
This PR adds a few concepts that allow IOPromises to be cancelled, or made inert, when a context/block scope is left. This is very similar to some of the concepts in golang's [`Context`](https://blog.golang.org/context). The diagrams in [this great post](https://www.sohamkamani.com/golang/context-cancellation-and-values/) explain the problem that Go was solving and that is the same problem that this PR aims to solve for IOPromise.

This PR is best viewed as the individual commits, which incrementally add this functionality.

Essentially this solves a potential concurrency risk introduced by IOPromises, which is that certain work can be batched up inside `then` handlers of IOPromise objects, which could reference shared scope, but may live beyond a request lifecycle by either missing a `sync` call or raising before one. In successful requests, this wouldn't present an issue since we would expect that all these promises would be `sync`ed at some point, ensuring that everything is cleaned up. However, if a request _fails_ due to an exception, it's possible that the `.sync` call would be bypassed and the promises could continue to live inside the IOPromise execution contexts. Since we expect the underlying storage connections to continue across requests, we can't simply throw these promises away.

This PR prevents these handlers from inadvertently executing late, by enabling every `IOPromise::Base` promise to be made inert by `cancel`ing it. It then creates a `CancelContext` around a block, which ensures that any promises that are still waiting on IO are at the very least prevented from executing promise chains, as would be the case for normal Promise instances that are not part of IOPromise. They would then be cleaned up by GC once IOPromise is done with them. There is also a Rack middleware that uses this functionality. 

The best place to see the problem and the way the context middleware fixes it is in this test: https://github.com/iopromise-ruby/iopromise/blob/c3d708db076ac90ada01e436d66d9e716c35c761/spec/rack/context_middleware_spec.rb

The best place to see that non-IOPromise chains are disabled as expected is in this test: https://github.com/iopromise-ruby/iopromise/commit/1f6e76c3064d739cd2d7151aeacd14c4400fc18e#diff-ef76ba7327c65542594766085cf6e813bec74b11fbf6517573d1948bfa34d48aR6

You can imagine a trivial but common example where this would happen, say inside a Rails view, an IOPromise to perform some slow query is begun, then an unrelated exception occurs before the call to `.sync`. Say this promise had a `then` handler with some more IO queries. This would leave the IOPromise executing, and eventually the `.then` handler would execute and potentially initiate another IO operation even though the request has already long finished. Instead, with this PR, in every case the IOPromise will be prevented from continuing.

Additionally, specific IOPromise implementations can optionally extend the `cancel` method in their subclass to actually cancel any downstream work that's pending - this is left up to individual implementations to decide in their own repos.

/cc @tma @mclark

theojulienne commented:
One thing to add that I forgot to explain: You might notice I added something here that nests `CancelContext` instances, even though this would never be possible right now because it's strictly block based. I think eventually it would be very useful to add some basic timeout support to these contexts (similar to Go), to ensure that no matter what happens the IOPromise loop would end after a certain amount of total time elapsed. I don't think this is as critical as the safety issue addressed here, but the design here allows a future path where nested contexts could potentially add timeouts and cancel early (and where the top-most context is potentially not the one to timeout and cancel).

mclark commented:
@theojulienne this is a hugely important thing to consider and without this notion of "cancelling" pending operations we could end up with some very serious consequences so thank you very much for diving in! 🙇 

When it comes to the topic data security, I prefer to keep things as simple as possible so that anyone can easily understand how the security is achieved. So when I think about leaking promise handling across requests, my default (naïve) approach would be to somehow reset the `ExecutorContext` for the thread so that the old one is garbage collected along with all the promises it references. Therefore, it wouldn't be able to process any "post IO" logic as it wouldn't exist in the first place. This wouldn't be as robust as a stack based implementation, but I don't think it would preclude that being added later either. Is there anything blocking us from doing something like this?

```ruby
module IOPromise
  class << self
    def with_executor_context
      @current = ExecutorContext.new
      yield
      @current = nil
    end

    attr_reader :current
  end
end
```

I'm not advocating for exactly this implementation (I don't know enough about this stuff yet to have a good solution) but something that makes it impossible for any further IOPromise related computation to occur.

Also it would be nice for promises to not need to know about the concept of being "cancelled" at all. Is there a way to `reject` all pending promises with an `IOPromise::CancelledError` or something before throwing away the `ExecutorContext`? 🤔

theojulienne commented:
Great questions!

> So when I think about leaking promise handling across requests, my default (naïve) approach would be to somehow reset the ExecutorContext for the thread so that the old one is garbage collected along with all the promises it references.

This is actually what I was initially planning, the problem that I found is that we do expect connections to last beyond a single request, and in IOPromise land, an ExecutorContext essentially manages the connection and how data is streamed to/from it. Because of this, if the ExecutorContext itself is thrown away, the state about what was currently running on the connection would as well, which would make the connection unsafe to reuse, because we could be part way through sending/receiving a request and would lose that state.

The pattern in this PR (which I'm not tied to, but was the best I've come up with so far) means that the connection and request/response logic is maintained and can complete (unless an underlying implementation explicitly has a way to cancel pending requests), but application level code is prevented from executing on the result further.

> Also it would be nice for promises to not need to know about the concept of being "cancelled" at all. Is there a way to reject all pending promises with an IOPromise::CancelledError or something before throwing away the ExecutorContext? 🤔

I was also initially thinking we could do `reject` rather than `cancel`, except rejections will execute application code and can contain scope and follow up actions, so are not safe either, consider something like this:
```ruby
def async_foo(item)
  foo = async_too_slow.rescue do
     async_some_other_way(item)
  end
end
```

If the promise behind `async_too_slow` times out past the end of the request, and was rejected, the rescue block would be executed creating a new promise or otherwise executing on the `item` scope variable, which it should not be able to.

The concept of "cancelling" does prevent this above example, though I'd be interested in any other ways we can make this safer as well 🤔

mclark commented:
This is great, thanks so much for all the context @theojulienne!

> if the ExecutorContext itself is thrown away, the state about what was currently running on the connection would as well, which would make the connection unsafe to reuse

This makes complete sense, I agree, let's not GC the `ExecutorContext` 😅 

> I was also initially thinking we could do reject rather than cancel, except rejections will execute application code

🤦 of course, I forgot all about that! Again, excellent point.

Looking through this again and the `iopromise-faraday` project, I am starting to see why we need to use the promise as the place to manage cancellations. For readers who may be as confused as I, other libraries that implement the parallel IO likely have callbacks for when their work is complete and it is fairly common to just fulfill the promise with that data. For example in [iopromise-faraday](https://github.com/iopromise-ruby/iopromise-faraday/blob/main/lib/iopromise/faraday/promise.rb#L20-L22). These libraries _may_ have support for cancelling but it's possible they don't so we need to have a guaranteed hard stop in the promise code.

New concerns based on my incrementally larger understanding of the problem space:
* We are overriding private `Promise` methods to achieve the goal of making them inert. What do you think of overriding public methods such as `fulfill` and `reject` instead?
* It looks like there is nothing stopping an IOPromise implementor from hooking in a regular `Promise` in to the chain when a parallel call returns. If that happens, I don't think the cancel context can really save us from anything. That being said, an implementor can really just call anything they like in these callbacks so maybe we just have to watch out to ensure that code is safe.

theojulienne commented:
> We are overriding private Promise methods to achieve the goal of making them inert. What do you think of overriding public methods such as fulfill and reject instead?

This seems very reasonable to me :+1: Trying to switch to this actually revealed some corner cases around pending promises being cancelled (if `sync` isn't called at all between creation and context cancellation) and cancellation needing to clean promises up from the pending promise list in the `Deferred` implementation. I've also added additional sanity checks to make sure we never wait on a cancelled promise, since it would never complete. I've addressed all of these and switched to overriding the public methods in 2a13447 - I think this is looking much better. I still wish it was simpler, though.

> It looks like there is nothing stopping an IOPromise implementor from hooking in a regular Promise in to the chain when a parallel call returns. If that happens, I don't think the cancel context can really save us from anything. That being said, an implementor can really just call anything they like in these callbacks so maybe we just have to watch out to ensure that code is safe.

Agreed, the implementations of the IOPromise subclasses do need to be cautious about what they do inside callbacks. Even if a promise chain is used, we would still expect the final "end user API" of the IOPromise to be blocked from running application code, so I think this is as good as we can guarantee. For the most part, the cancellation is an internal API that application level users shouldn't need to think about.

mclark commented on lib/iopromise.rb:
```diff
@@ -41,13 +43,41 @@ def notify_completion(value: nil, reason: nil)
     end
 
     def fulfill(value)
+      return if defined?(@cancelled)
```
```suggestion
      return if cancelled?
```

mclark commented on lib/iopromise.rb:
```diff
@@ -41,13 +43,41 @@ def notify_completion(value: nil, reason: nil)
     end
 
     def reject(reason)
+      return if defined?(@cancelled)
```
```suggestion
      return if cancelled?
```

mclark commented on spec/base_spec.rb:
```diff
@@ -0,0 +1,97 @@
+
+      b.cancel
+
+      expect(b).to_not be_pending
```
What happens if I do something like this after cancelling `b`?

```ruby
b.then { |v| v }
```

Does the chained promise execute? Should we override `then` as well? I might be getting a bit too paranoid with this suggestion 😄

theojulienne replied:
In the current implementation, cancelling does nothing once a promise is resolved, which I think is acceptable since it's not a time delayed execution from the shared IOPromise event loop at that point. I believe the only time there's a risk is where a promise ends up pending but not synced, inside the IOPromise event loop.

theojulienne replied:
I'm going to add a test for `.then` in the case where cancellation happens (before resolving) and the case where cancellation doesn't do anything (after resolving), which makes this behaviour explicit in the form of a test. I feel like changing `cancel` to act on already-resolved promises could get a bit more complex (since actions up to that point already completed) but if there's a case where this would prevent an issue it might make sense?

mclark commented on spec/cancel_context_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      end
+
+      expect(deferred2).to be_pending
+      expect(deferred2).to be_cancelled
```
```suggestion
      expect(deferred2).to be_cancelled
      expect(deferred).to_not be_cancelled
```

mclark reviewed (commented):
Just some minor suggestions, I think this is good to merge 👍

mclark reviewed (approved):
🚢

BlakeWilliams commented on spec/cancel_context_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    expect(deferred).to be_pending
+    expect(deferred).to be_cancelled
+    expect(chained).to be_pending
+    expect(chained).to_not be_cancelled
```
I don't have a ton of context, but when reading through this test case I thought that `chained` would have been canceled since its parent, `deferred`, was canceled. I'm curious if that's an intentional design decision or a side effect of the promise implementation?

theojulienne replied:
That's correct, we don't propagate cancellation down the chain, and the promises generated with `.then {}` are never registered as IOPromises under the current execution/main loop, so never have the potential to be triggered outside of the request.

Of course, some other code can still trigger these to resolve since they aren't cancelled, but that goes for any `Promise`, and it would be quite a lot of extra housekeeping to register all pending promises in the chain in order to know about them and cancel them. Because of this, we just prevent the IO-based promises themselves from being acted upon once a request completes. In theory, this should provide the same safety guarantee as normal promises.
